cmd/juju/caas/add.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caas
+
+import (
+	"fmt"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+
+	"github.com/juju/juju/api/base"
+	cloudapi "github.com/juju/juju/api/cloud"
+	"github.com/juju/juju/cloud"
+	"github.com/juju/juju/cmd/juju/common"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/jujuclient"
+)
+
+type CloudMetadataStore interface {
+	ParseCloudMetadataFile(path string) (map[string]cloud.Cloud, error)
+	ParseOneCloud(data []byte) (cloud.Cloud, error)
+	PublicCloudMetadata(searchPaths ...string) (result map[string]cloud.Cloud, fallbackUsed bool, _ error)
+	PersonalCloudMetadata() (map[string]cloud.Cloud, error)
+	WritePersonalCloudMetadata(cloudsMap map[string]cloud.Cloud) error
+}
+
+// Implemented by cloudapi.Client
+type AddCloudAPI interface {
+	AddCloud(cloud.Cloud) error
+}
+
+var usageAddCaasSummary = `
```

Comment:
⚠️  jsing pointed out that this isn't proper go naming. Needs to be usageAddCAASSummary. Lots of other examples in the branch, will fix.